File: app/src/test/java/it/cosenonjaviste/core/CnjJUnitDaggerRule.java
package it.cosenonjaviste.core;

import org.junit.runners.model.FrameworkMethod;
import org.junit.runners.model.Statement;

import io.reactivex.android.plugins.RxAndroidPlugins;
import io.reactivex.plugins.RxJavaPlugins;
import io.reactivex.schedulers.Schedulers;
import it.cosenonjaviste.daggermock.DaggerMockRule;
import it.cosenonjaviste.model.TwitterService;
import it.cosenonjaviste.model.WordPressService;
import it.cosenonjaviste.ui.AppModule;
import it.cosenonjaviste.ui.ApplicationComponent;

public class CnjJUnitDaggerRule extends DaggerMockRule<ApplicationComponent> {
    public CnjJUnitDaggerRule() {
        super(ApplicationComponent.class, new AppModule(null));
        providesMock(WordPressService.class, TwitterService.class);
    }

    @Override public Statement apply(Statement base, FrameworkMethod method, Object target) {
        Statement superStatement = super.apply(base, method, target);
        return new Statement() {
            @Override public void evaluate() throws Throwable {
                RxJavaPlugins.setIoSchedulerHandler(
                        scheduler -> Schedulers.trampoline());
                RxJavaPlugins.setComputationSchedulerHandler(
                        scheduler -> Schedulers.trampoline());
                RxJavaPlugins.setNewThreadSchedulerHandler(
                        scheduler -> Schedulers.trampoline());
                RxAndroidPlugins.setInitMainThreadSchedulerHandler(
                        scheduler -> Schedulers.trampoline());

                try
                {
                    superStatement.evaluate();
                } finally
                {
                    RxJavaPlugins.reset();
                    RxAndroidPlugins.reset();
                }
            }
        };
    }
}

File: app/src/test/java/it/cosenonjaviste/core/CnjJUnitDaggerRule.kt
package it.cosenonjaviste.core

import it.cosenonjaviste.daggermock.DaggerMockRule
import it.cosenonjaviste.model.TwitterService
import it.cosenonjaviste.model.WordPressService
import it.cosenonjaviste.ui.AppModule
import it.cosenonjaviste.ui.ApplicationComponent
import org.junit.runners.model.FrameworkMethod
import org.junit.runners.model.Statement
import rx.Scheduler
import rx.android.plugins.RxAndroidPlugins
import rx.android.plugins.RxAndroidSchedulersHook
import rx.plugins.RxJavaHooks
import rx.schedulers.Schedulers

class CnjJUnitDaggerRule : DaggerMockRule<ApplicationComponent>(ApplicationComponent::class.java, AppModule(null)) {
    init {
        providesMock(WordPressService::class.java, TwitterService::class.java)
    }

    override fun apply(base: Statement, method: FrameworkMethod?, target: Any): Statement {
        val superStatement = super.apply(base, method, target)
        return object : Statement() {
            @Throws(Throwable::class)
            override fun evaluate() {
                RxJavaHooks.setOnIOScheduler { scheduler -> Schedulers.immediate() }
                RxJavaHooks.setOnComputationScheduler { scheduler -> Schedulers.immediate() }
                RxJavaHooks.setOnNewThreadScheduler { scheduler -> Schedulers.immediate() }
                RxAndroidPlugins.getInstance().registerSchedulersHook(object : RxAndroidSchedulersHook() {
                    override fun getMainThreadScheduler(): Scheduler {
                        return Schedulers.immediate()
                    }
                })
                try {
                    superStatement.evaluate()
                } finally {
                    RxJavaHooks.reset()
                    RxAndroidPlugins.getInstance().reset()
                }
            }
        }
    }
}
